// phosphor-dockpanel/lib/index.js
/*-----------------------------------------------------------------------------
| Copyright (c) 2014-2015, PhosphorJS Contributors
|
| Distributed under the terms of the BSD 3-Clause License.
|
| The full license is in the file LICENSE, distributed with this software.
|----------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var arrays = require('phosphor-arrays');
var phosphor_domutil_1 = require('phosphor-domutil');
var phosphor_dragdrop_1 = require('phosphor-dragdrop');
var phosphor_nodewrapper_1 = require('phosphor-nodewrapper');
var phosphor_properties_1 = require('phosphor-properties');
var phosphor_splitpanel_1 = require('phosphor-splitpanel');
var phosphor_stackedpanel_1 = require('phosphor-stackedpanel');
var phosphor_tabs_1 = require('phosphor-tabs');
var phosphor_widget_1 = require('phosphor-widget');
require('./index.css');
// TODO - need better solution for storing these class names
/**
 * The class name added to DockPanel instances.
 */
var DOCK_PANEL_CLASS = 'p-DockPanel';
/**
 * The class name added to dock tab panels.
 */
var TAB_PANEL_CLASS = 'p-DockTabPanel';
/**
 * The class name added to dock split panels.
 */
var SPLIT_PANEL_CLASS = 'p-DockSplitPanel';
/**
 * The class name added to dock panel overlays.
 */
var OVERLAY_CLASS = 'p-DockPanel-overlay';
/**
 * The class name added to hidden overlays and tabs.
 */
var HIDDEN_CLASS = 'p-mod-hidden';
/**
 * The class name added to top root dock overlays.
 */
var ROOT_TOP_CLASS = 'p-mod-root-top';
/**
 * The class name added to left root dock overlays.
 */
var ROOT_LEFT_CLASS = 'p-mod-root-left';
/**
 * The class name added to right root dock overlays.
 */
var ROOT_RIGHT_CLASS = 'p-mod-root-right';
/**
 * The class name added to bottom root dock overlays.
 */
var ROOT_BOTTOM_CLASS = 'p-mod-root-bottom';
/**
 * The class name added to center root dock overlays.
 */
var ROOT_CENTER_CLASS = 'p-mod-root-center';
/**
 * The class name added to top panel dock overlays.
 */
var PANEL_TOP_CLASS = 'p-mod-panel-top';
/**
 * The class name added to left panel dock overlays.
 */
var PANEL_LEFT_CLASS = 'p-mod-panel-left';
/**
 * The class name added to right panel dock overlays.
 */
var PANEL_RIGHT_CLASS = 'p-mod-panel-right';
/**
 * The class name added to bottom panel dock overlays.
 */
var PANEL_BOTTOM_CLASS = 'p-mod-panel-bottom';
/**
 * The class named added to center panel dock overlays.
 */
var PANEL_CENTER_CLASS = 'p-mod-panel-center';
/**
 * The factory MIME type supported by the dock panel.
 */
var FACTORY_MIME = 'application/x-phosphor-widget-factory';
/**
 * The size of the edge dock zone for the root panel.
 */
var EDGE_SIZE = 30;
/**
 * A widget which provides a flexible docking area for content widgets.
 */
var DockPanel = (function (_super) {
    __extends(DockPanel, _super);
    /**
     * Construct a new dock panel.
     */
    function DockPanel() {
        _super.call(this);
        this.addClass(DOCK_PANEL_CLASS);
        this.layout = new phosphor_stackedpanel_1.StackedLayout();
    }
    Object.defineProperty(DockPanel.prototype, "spacing", {
        /**
         * Get the spacing between the tab panels.
         */
        get: function () {
            return DockPanelPrivate.spacingProperty.get(this);
        },
        /**
         * Set the spacing between the tab panels.
         */
        set: function (value) {
            DockPanelPrivate.spacingProperty.set(this, value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Insert a widget as a new panel above a reference widget.
     *
     * @param widget - The widget to insert into the dock panel.
     *
     * @param ref - The reference widget. If this is not provided, the
     *   widget will be inserted at the top edge of the dock panel.
     *
     * @throws An error if either `widget` or `ref` is invalid.
     */
    DockPanel.prototype.insertTop = function (widget, ref) {
        DockPanelPrivate.insertSplit(this, widget, ref, phosphor_splitpanel_1.Orientation.Vertical, false);
    };
    /**
     * Insert a widget as a new panel to the left of a reference widget.
     *
     * @param widget - The widget to insert into the dock panel.
     *
     * @param ref - The reference widget. If this is not provided, the
     *   widget will be inserted at the left edge of the dock panel.
     *
     * @throws An error if either `widget` or `ref` is invalid.
     */
    DockPanel.prototype.insertLeft = function (widget, ref) {
        DockPanelPrivate.insertSplit(this, widget, ref, phosphor_splitpanel_1.Orientation.Horizontal, false);
    };
    /**
     * Insert a widget as a new panel to the right of a reference widget.
     *
     * @param widget - The widget to insert into the dock panel.
     *
     * @param ref - The reference widget. If this is not provided, the
     *   widget will be inserted at the right edge of the dock panel.
     *
     * @throws An error if either `widget` or `ref` is invalid.
     */
    DockPanel.prototype.insertRight = function (widget, ref) {
        DockPanelPrivate.insertSplit(this, widget, ref, phosphor_splitpanel_1.Orientation.Horizontal, true);
    };
    /**
     * Insert a widget as a new panel below a reference widget.
     *
     * @param widget - The widget to insert into the dock panel.
     *
     * @param ref - The reference widget. If this is not provided, the
     *   widget will be inserted at the bottom edge of the dock panel.
     *
     * @throws An error if either `widget` or `ref` is invalid.
     */
    DockPanel.prototype.insertBottom = function (widget, ref) {
        DockPanelPrivate.insertSplit(this, widget, ref, phosphor_splitpanel_1.Orientation.Vertical, true);
    };
    /**
     * Insert a widget as a sibling tab before a reference widget.
     *
     * @param widget - The widget to insert into the dock panel.
     *
     * @param ref - The reference widget. If this is not provided, the
     *   widget will be inserted as the first tab in the top-left panel.
     *
     * @throws An error if either `widget` or `ref` is invalid.
     */
    DockPanel.prototype.insertTabBefore = function (widget, ref) {
        DockPanelPrivate.insertTab(this, widget, ref, false);
    };
    /**
     * Insert a widget as a sibling tab after a reference widget.
     *
     * @param widget - The widget to insert into the dock panel.
     *
     * @param ref - The reference widget. If this is not provided, the
     *   widget will be inserted as the last tab in the top-left panel.
     *
     * @throws An error if either `widget` or `ref` is invalid.
     */
    DockPanel.prototype.insertTabAfter = function (widget, ref) {
        DockPanelPrivate.insertTab(this, widget, ref, true);
    };
    /**
     * Ensure the tab for the specified content widget is selected.
     *
     * @param widget - The content widget of interest.
     *
     * #### Notes
     * If the widget is not contained in the dock panel, or is already
     * the selected tab in its respective tab panel, this is a no-op.
     */
    DockPanel.prototype.selectWidget = function (widget) {
        DockPanelPrivate.selectWidget(this, widget);
    };
    /**
     * Handle the DOM events for the dock panel.
     *
     * @param event - The DOM event sent to the dock panel.
     *
     * #### Notes
     * This method implements the DOM `EventListener` interface and is
     * called in response to events on the dock panel's node. It should
     * not be called directly by user code.
     */
    DockPanel.prototype.handleEvent = function (event) {
        switch (event.type) {
            case 'p-dragenter':
                this._evtDragEnter(event);
                break;
            case 'p-dragleave':
                this._evtDragLeave(event);
                break;
            case 'p-dragover':
                this._evtDragOver(event);
                break;
            case 'p-drop':
                this._evtDrop(event);
                break;
        }
    };
    /**
     * A message handler invoked on an `'after-attach'` message.
     */
    DockPanel.prototype.onAfterAttach = function (msg) {
        var node = this.node;
        node.addEventListener('p-dragenter', this);
        node.addEventListener('p-dragleave', this);
        node.addEventListener('p-dragover', this);
        node.addEventListener('p-drop', this);
    };
    /**
     * A message handler invoked on a `'before-detach'` message.
     */
    DockPanel.prototype.onBeforeDetach = function (msg) {
        var node = this.node;
        node.removeEventListener('p-dragenter', this);
        node.removeEventListener('p-dragleave', this);
        node.removeEventListener('p-dragover', this);
        node.removeEventListener('p-drop', this);
    };
    /**
     * Handle the `'p-dragenter'` event for the dock panel.
     */
    DockPanel.prototype._evtDragEnter = function (event) {
        if (event.mimeData.hasData(FACTORY_MIME)) {
            event.preventDefault();
            event.stopPropagation();
        }
    };
    /**
     * Handle the `'p-dragleave'` event for the dock panel.
     */
    DockPanel.prototype._evtDragLeave = function (event) {
        event.preventDefault();
        event.stopPropagation();
        var related = event.relatedTarget;
        if (!related || !this.node.contains(related)) {
            DockPanelPrivate.hideOverlay(this);
        }
    };
    /**
     * Handle the `'p-dragover'` event for the dock panel.
     */
    DockPanel.prototype._evtDragOver = function (event) {
        event.preventDefault();
        event.stopPropagation();
        var x = event.clientX;
        var y = event.clientY;
        var zone = DockPanelPrivate.showOverlay(this, x, y);
        if (zone === 10 /* Invalid */) {
            event.dropAction = phosphor_dragdrop_1.DropAction.None;
        }
        else {
            event.dropAction = event.proposedAction;
        }
    };
    /**
     * Handle the `'p-drop'` event for the dock panel.
     */
    DockPanel.prototype._evtDrop = function (event) {
        event.preventDefault();
        event.stopPropagation();
        DockPanelPrivate.hideOverlay(this);
        if (event.proposedAction === phosphor_dragdrop_1.DropAction.None) {
            event.dropAction = phosphor_dragdrop_1.DropAction.None;
            return;
        }
        var x = event.clientX;
        var y = event.clientY;
        var target = DockPanelPrivate.findDockTarget(this, x, y);
        if (target.zone === 10 /* Invalid */) {
            event.dropAction = phosphor_dragdrop_1.DropAction.None;
            return;
        }
        var factory = event.mimeData.getData(FACTORY_MIME);
        if (typeof factory !== 'function') {
            event.dropAction = phosphor_dragdrop_1.DropAction.None;
            return;
        }
        var widget = factory();
        if (!(widget instanceof phosphor_widget_1.Widget)) {
            event.dropAction = phosphor_dragdrop_1.DropAction.None;
            return;
        }
        DockPanelPrivate.handleDrop(this, widget, target);
        event.dropAction = event.proposedAction;
    };
    return DockPanel;
})(phosphor_widget_1.Widget);
exports.DockPanel = DockPanel;
/**
 * A custom tab panel used by a DockPanel.
 */
var DockTabPanel = (function (_super) {
    __extends(DockTabPanel, _super);
    /**
     * Construct a new dock tab panel.
     */
    function DockTabPanel() {
        _super.call(this);
        this.addClass(TAB_PANEL_CLASS);
        this.tabBar.tabsMovable = true;
    }
    return DockTabPanel;
})(phosphor_tabs_1.TabPanel);
/**
 * A custom split panel used by a DockPanel.
 */
var DockSplitPanel = (function (_super) {
    __extends(DockSplitPanel, _super);
    /**
     * Construct a new dock split panel.
     */
    function DockSplitPanel(orientation, spacing) {
        _super.call(this);
        this.addClass(SPLIT_PANEL_CLASS);
        this.orientation = orientation;
        this.spacing = spacing;
    }
    return DockSplitPanel;
})(phosphor_splitpanel_1.SplitPanel);
/**
 * A node wrapper used as an overlay dock indicator for a dock panel.
 */
var DockPanelOverlay = (function (_super) {
    __extends(DockPanelOverlay, _super);
    /**
     * Construct a new dock panel overlay.
     */
    function DockPanelOverlay() {
        _super.call(this);
        this._zone = 10 /* Invalid */;
        this.addClass(OVERLAY_CLASS);
        this.addClass(HIDDEN_CLASS);
    }
    /**
     * Show the overlay with the given zone and geometry
     */
    DockPanelOverlay.prototype.show = function (zone, left, top, width, height) {
        var style = this.node.style;
        style.top = top + 'px';
        style.left = left + 'px';
        style.width = width + 'px';
        style.height = height + 'px';
        this.removeClass(HIDDEN_CLASS);
        this._setZone(zone);
    };
    /**
     * Hide the overlay and reset its zone.
     */
    DockPanelOverlay.prototype.hide = function () {
        this.addClass(HIDDEN_CLASS);
        this._setZone(10 /* Invalid */);
    };
    /**
     * Set the dock zone for the overlay.
     */
    DockPanelOverlay.prototype._setZone = function (zone) {
        if (zone === this._zone) {
            return;
        }
        var oldClass = DockPanelOverlay.zoneMap[this._zone];
        var newClass = DockPanelOverlay.zoneMap[zone];
        if (oldClass)
            this.removeClass(oldClass);
        if (newClass)
            this.addClass(newClass);
        this._zone = zone;
    };
    /**
     * A mapping of dock zone enum value to modifier class.
     */
    DockPanelOverlay.zoneMap = [
        ROOT_TOP_CLASS,
        ROOT_LEFT_CLASS,
        ROOT_RIGHT_CLASS,
        ROOT_BOTTOM_CLASS,
        ROOT_CENTER_CLASS,
        PANEL_TOP_CLASS,
        PANEL_LEFT_CLASS,
        PANEL_RIGHT_CLASS,
        PANEL_BOTTOM_CLASS,
        PANEL_CENTER_CLASS
    ];
    return DockPanelOverlay;
})(phosphor_nodewrapper_1.NodeWrapper);
/**
 * The namespace for the `DockPanel` class private data.
 */
var DockPanelPrivate;
(function (DockPanelPrivate) {
    /**
     * The property descriptor for the spacing between panels.
     */
    DockPanelPrivate.spacingProperty = new phosphor_properties_1.Property({
        name: 'spacing',
        value: 3,
        coerce: function (owner, value) { return Math.max(0, value | 0); },
        changed: onSpacingChanged,
    });
    /**
     * Insert a widget as a new split panel in a dock panel.
     */
    function insertSplit(owner, widget, ref, orientation, after) {
        // Ensure the insert args are valid.
        validateInsertArgs(owner, widget, ref);
        // If the widget is the same as the ref, there's nothing to do.
        if (widget === ref) {
            return;
        }
        // Unparent the widget before performing the insert. This ensures
        // that structural changes to the dock panel occur before searching
        // for the insert location.
        widget.parent = null;
        // Setup the new tab panel to host the widget.
        var tabPanel = createTabPanel();
        tabPanel.addChild(widget);
        // If there is no root, add the new tab panel as the root.
        if (!getRoot(owner)) {
            setRoot(owner, tabPanel);
            return;
        }
        // If the ref widget is null, split the root panel.
        if (!ref) {
            var root = ensureSplitRoot(owner, orientation);
            var sizes_1 = root.sizes();
            var count = sizes_1.length;
            arrays.insert(sizes_1, after ? count : 0, 0.5);
            root.insertChild(after ? count : 0, tabPanel);
            root.setSizes(sizes_1);
            return;
        }
        // Lookup the tab panel for the ref widget.
        var refTabPanel = findTabPanel(ref);
        // If the ref tab panel parent is the dock panel, split the root.
        if (refTabPanel.parent === owner) {
            var root = ensureSplitRoot(owner, orientation);
            root.insertChild(after ? 1 : 0, tabPanel);
            root.setSizes([1, 1]);
            return;
        }
        // Assert the parent of the ref tab panel is a dock split panel.
        if (!(refTabPanel.parent instanceof DockSplitPanel)) {
            internalError();
        }
        // Cast the ref tab panel parent to a dock split panel.
        var splitPanel = refTabPanel.parent;
        // If the split panel is the correct orientation, the widget
        // can be inserted directly and sized to 1/2 the ref space.
        if (splitPanel.orientation === orientation) {
            var i_1 = splitPanel.childIndex(refTabPanel);
            var sizes_2 = splitPanel.sizes();
            var size = sizes_2[i_1] = sizes_2[i_1] / 2;
            arrays.insert(sizes_2, after ? i_1 + 1 : i_1, size);
            splitPanel.insertChild(after ? i_1 + 1 : i_1, tabPanel);
            splitPanel.setSizes(sizes_2);
            return;
        }
        // If the split panel only has a single child, its orientation
        // can be changed directly and its sizes set to a 1:1 ratio.
        if (splitPanel.childCount() === 1) {
            splitPanel.orientation = orientation;
            splitPanel.insertChild(after ? 1 : 0, tabPanel);
            splitPanel.setSizes([1, 1]);
            return;
        }
        // Assert the split panel has more than one child.
        if (splitPanel.childCount() === 0) {
            internalError();
        }
        // Otherwise, a new split panel with the correct orientation needs
        // to be created to hold the ref panel and tab panel, and inserted
        // in the previous location of the ref panel.
        var sizes = splitPanel.sizes();
        var i = splitPanel.childIndex(refTabPanel);
        var childSplit = new DockSplitPanel(orientation, owner.spacing);
        childSplit.addChild(refTabPanel);
        childSplit.insertChild(after ? 1 : 0, tabPanel);
        splitPanel.insertChild(i, childSplit);
        splitPanel.setSizes(sizes);
        childSplit.setSizes([1, 1]);
    }
    DockPanelPrivate.insertSplit = insertSplit;
    /**
     * Insert a widget as a sibling tab in a dock panel.
     */
    function insertTab(owner, widget, ref, after) {
        // Ensure the insert args are valid.
        validateInsertArgs(owner, widget, ref);
        // If the widget is the same as the ref, there's nothing to do.
        if (widget === ref) {
            return;
        }
        // Unparent the widget before performing the insert. This ensures
        // that structural changes to the dock panel occur before searching
        // for the insert location.
        widget.parent = null;
        // Find the index and tab panel for the insert operation.
        var index;
        var tabPanel;
        if (ref) {
            tabPanel = findTabPanel(ref);
            index = tabPanel.childIndex(ref) + (after ? 1 : 0);
        }
        else {
            tabPanel = ensureFirstTabPanel(owner);
            index = after ? tabPanel.childCount() : 0;
        }
        // Insert the widget into the tab panel at the proper location.
        tabPanel.insertChild(index, widget);
    }
    DockPanelPrivate.insertTab = insertTab;
    /**
     * Ensure the given widget is the current widget in its tab panel.
     *
     * This is a no-op if the widget is not contained in the dock panel.
     */
    function selectWidget(owner, widget) {
        if (!dockPanelContains(owner, widget))
            return;
        widget.parent.parent.currentWidget = widget;
    }
    DockPanelPrivate.selectWidget = selectWidget;
    /**
     * Hide the dock panel overlay for the given dock panel.
     */
    function hideOverlay(owner) {
        getOverlay(owner).hide();
    }
    DockPanelPrivate.hideOverlay = hideOverlay;
    /**
     * Show the dock panel overlay indicator at the given client position.
     *
     * If the position is not over a dock zone, the overlay is hidden.
     *
     * This returns the dock zone used to display the overlay.
     */
    function showOverlay(owner, clientX, clientY) {
        // Find the dock target for the given client position.
        var target = findDockTarget(owner, clientX, clientY);
        // If the dock zone is invalid, hide the overlay and bail.
        if (target.zone === 10 /* Invalid */) {
            hideOverlay(owner);
            return target.zone;
        }
        // Setup the variables needed to compute the overlay geometry.
        var top;
        var left;
        var width;
        var height;
        var pcr;
        var box = phosphor_domutil_1.boxSizing(owner.node); // TODO cache this?
        var rect = owner.node.getBoundingClientRect();
        // Compute the overlay geometry based on the dock zone.
        switch (target.zone) {
            case 0 /* RootTop */:
                top = box.paddingTop;
                left = box.paddingLeft;
                width = rect.width - box.horizontalSum;
                height = (rect.height - box.verticalSum) / 3;
                break;
            case 1 /* RootLeft */:
                top = box.paddingTop;
                left = box.paddingLeft;
                width = (rect.width - box.horizontalSum) / 3;
                height = rect.height - box.verticalSum;
                break;
            case 2 /* RootRight */:
                top = box.paddingTop;
                width = (rect.width - box.horizontalSum) / 3;
                left = box.paddingLeft + 2 * width;
                height = rect.height - box.verticalSum;
                break;
            case 3 /* RootBottom */:
                height = (rect.height - box.verticalSum) / 3;
                top = box.paddingTop + 2 * height;
                left = box.paddingLeft;
                width = rect.width - box.horizontalSum;
                break;
            case 4 /* RootCenter */:
                top = box.paddingTop;
                left = box.paddingLeft;
                width = rect.width - box.horizontalSum;
                height = rect.height - box.verticalSum;
                break;
            case 5 /* PanelTop */:
                pcr = target.panel.node.getBoundingClientRect();
                top = pcr.top - rect.top - box.borderTop;
                left = pcr.left - rect.left - box.borderLeft;
                width = pcr.width;
                height = pcr.height / 2;
                break;
            case 6 /* PanelLeft */:
                pcr = target.panel.node.getBoundingClientRect();
                top = pcr.top - rect.top - box.borderTop;
                left = pcr.left - rect.left - box.borderLeft;
                width = pcr.width / 2;
                height = pcr.height;
                break;
            case 7 /* PanelRight */:
                pcr = target.panel.node.getBoundingClientRect();
                top = pcr.top - rect.top - box.borderTop;
                left = pcr.left - rect.left - box.borderLeft + pcr.width / 2;
                width = pcr.width / 2;
                height = pcr.height;
                break;
            case 8 /* PanelBottom */:
                pcr = target.panel.node.getBoundingClientRect();
                top = pcr.top - rect.top - box.borderTop + pcr.height / 2;
                left = pcr.left - rect.left - box.borderLeft;
                width = pcr.width;
                height = pcr.height / 2;
                break;
            case 9 /* PanelCenter */:
                pcr = target.panel.node.getBoundingClientRect();
                top = pcr.top - rect.top - box.borderTop;
                left = pcr.left - rect.left - box.borderLeft;
                width = pcr.width;
                height = pcr.height;
                break;
        }
        // Show the overlay and return the dock zone.
        getOverlay(owner).show(target.zone, left, top, width, height);
        return target.zone;
    }
    DockPanelPrivate.showOverlay = showOverlay;
    /**
     * Find the dock target for the given client position.
     */
    function findDockTarget(owner, clientX, clientY) {
        var root = getRoot(owner);
        if (!root) {
            return { zone: 4 /* RootCenter */, panel: null };
        }
        if (!phosphor_domutil_1.hitTest(root.node, clientX, clientY)) {
            return { zone: 10 /* Invalid */, panel: null };
        }
        var edgeZone = getEdgeZone(root.node, clientX, clientY);
        if (edgeZone !== 10 /* Invalid */) {
            return { zone: edgeZone, panel: null };
        }
        var hitPanel = iterTabPanels(root, function (tabs) {
            return phosphor_domutil_1.hitTest(tabs.node, clientX, clientY) ? tabs : void 0;
        });
        if (!hitPanel) {
            return { zone: 10 /* Invalid */, panel: null };
        }
        var panelZone = getPanelZone(hitPanel.node, clientX, clientY);
        return { zone: panelZone, panel: hitPanel };
    }
    DockPanelPrivate.findDockTarget = findDockTarget;
    /**
     * Drop a widget onto a dock panel using the given dock target.
     */
    function handleDrop(owner, widget, target) {
        // Do nothing if the dock zone is invalid.
        if (target.zone === 10 /* Invalid */) {
            return;
        }
        // Handle the simple case of root drops first.
        switch (target.zone) {
            case 0 /* RootTop */:
                owner.insertTop(widget);
                return;
            case 1 /* RootLeft */:
                owner.insertLeft(widget);
                return;
            case 2 /* RootRight */:
                owner.insertRight(widget);
                return;
            case 3 /* RootBottom */:
                owner.insertBottom(widget);
                return;
            case 4 /* RootCenter */:
                owner.insertLeft(widget);
                return;
        }
        // Otherwise, it's a panel drop, and that requires more checks.
        // Do nothing if the widget is dropped as a tab on its own panel.
        if (target.zone === 9 /* PanelCenter */) {
            if (target.panel.childIndex(widget) !== -1) {
                return;
            }
        }
        // Do nothing if the panel only contains the drop widget.
        if (target.panel.childCount() === 1) {
            if (target.panel.childAt(0) === widget) {
                return;
            }
        }
        // Find a suitable reference widget for the drop.
        var n = target.panel.childCount();
        var ref = target.panel.childAt(n - 1);
        if (ref === widget) {
            ref = target.panel.childAt(n - 2);
        }
        // Insert the widget based on the panel zone.
        switch (target.zone) {
            case 5 /* PanelTop */:
                owner.insertTop(widget, ref);
                return;
            case 6 /* PanelLeft */:
                owner.insertLeft(widget, ref);
                return;
            case 7 /* PanelRight */:
                owner.insertRight(widget, ref);
                return;
            case 8 /* PanelBottom */:
                owner.insertBottom(widget, ref);
                return;
            case 9 /* PanelCenter */:
                owner.insertTabAfter(widget, ref);
                selectWidget(owner, widget);
                return;
        }
    }
    DockPanelPrivate.handleDrop = handleDrop;
    /**
     * A private attached property for the dock panel root.
     */
    var rootProperty = new phosphor_properties_1.Property({
        name: 'root',
        value: null,
        changed: onRootChanged,
    });
    /**
     * A private attached property for the dock panel overlay.
     */
    var overlayProperty = new phosphor_properties_1.Property({
        name: 'overlay',
        create: createOverlay,
    });
    /**
     * Get the root panel for a dock panel.
     */
    function getRoot(owner) {
        return rootProperty.get(owner);
    }
    /**
     * Set the root panel for a dock panel.
     */
    function setRoot(owner, root) {
        rootProperty.set(owner, root);
    }
    /**
     * Get the overlay for a dock panel.
     */
    function getOverlay(owner) {
        return overlayProperty.get(owner);
    }
    /**
     * The change handler for the dock panel `rootProperty`.
     *
     * This will re-parent the new root and set it as the current widget.
     *
     * The old root is not modified.
     */
    function onRootChanged(owner, old, root) {
        if (!root)
            return;
        var layout = owner.layout;
        layout.addChild(root);
        root.show();
    }
    /**
     * The creation handler for the dock panel `overlayProperty`.
     *
     * This will create and install the overlay for the panel.
     */
    function createOverlay(owner) {
        var overlay = new DockPanelOverlay();
        owner.node.appendChild(overlay.node);
        return overlay;
    }
    /**
     * The change handler for the `spacing` property of a dock panel.
     */
    function onSpacingChanged(owner, old, spacing) {
        var root = getRoot(owner);
        if (root instanceof DockSplitPanel) {
            updateSpacing(root, spacing);
        }
    }
    /**
     * Recursively update the spacing of a dock split panel.
     */
    function updateSpacing(panel, spacing) {
        for (var i = 0, n = panel.childCount(); i < n; ++i) {
            var child = panel.childAt(i);
            if (child instanceof DockSplitPanel) {
                updateSpacing(child, spacing);
            }
        }
        panel.spacing = spacing;
    }
    /**
     * Throw an internal dock panel error.
     */
    function internalError() {
        throw new Error('Internal DockPanel Error.');
    }
    /**
     * Test whether a dock panel contains the given widget.
     *
     * For this condition to be `true`, the widget must be a logical child
     * of a `DockTabPanel`, which itself must be a proper descendant of the
     * given dock panel.
     */
    function dockPanelContains(owner, widget) {
        var stack = widget.parent;
        if (!stack) {
            return false;
        }
        var tabs = stack.parent;
        if (!(tabs instanceof DockTabPanel)) {
            return false;
        }
        var parent = tabs.parent;
        while (parent) {
            if (parent === owner) {
                return true;
            }
            if (!(parent instanceof DockSplitPanel)) {
                return false;
            }
            parent = parent.parent;
        }
        return false;
    }
    /**
     * Find the ancestor dock tab panel for the given widget.
     *
     * This assumes the widget already belongs to a dock panel, and will
     * throw an error if that assumption does not hold.
     */
    function findTabPanel(widget) {
        var stack = widget.parent;
        if (!stack) {
            internalError();
        }
        var tabs = stack.parent;
        if (!(tabs instanceof DockTabPanel)) {
            internalError();
        }
        return tabs;
    }
    /**
     * Find the first dock tab panel for the given dock panel.
     *
     * This returns `null` if the dock panel has no content. It will throw
     * an error if the structure of the dock panel is found to be invalid.
     */
    function findFirstTabPanel(owner) {
        var root = getRoot(owner);
        while (root) {
            if (root instanceof DockTabPanel) {
                return root;
            }
            if (!(root instanceof DockSplitPanel) || root.childCount() === 0) {
                internalError();
            }
            root = root.childAt(0);
        }
        return null;
    }
    /**
     * Get or create the first dock tab panel for the given dock panel.
     *
     * If dock panel has no root, a new tab panel will be created and
     * added as the root. An error will be thrown if the structure of
     * the dock panel is found to be invalid.
     */
    function ensureFirstTabPanel(owner) {
        var tabs = findFirstTabPanel(owner);
        if (!tabs) {
            tabs = createTabPanel();
            setRoot(owner, tabs);
        }
        return tabs;
    }
    /**
     * Ensure the root panel is a splitter with the given orientation.
     *
     * This will throw an error if the panel does not have a current root,
     * since that would violate the invariants of the dock panel structure.
     */
    function ensureSplitRoot(owner, orientation) {
        var oldRoot = getRoot(owner);
        if (!oldRoot) {
            internalError();
        }
        if (oldRoot instanceof DockSplitPanel) {
            if (oldRoot.orientation === orientation) {
                return oldRoot;
            }
            if (oldRoot.childCount() <= 1) {
                oldRoot.orientation = orientation;
                return oldRoot;
            }
        }
        var newRoot = new DockSplitPanel(orientation, owner.spacing);
        newRoot.addChild(oldRoot);
        setRoot(owner, newRoot);
        return newRoot;
    }
    /**
     * Validate the insert arguments for a dock panel.
     *
     * This will throw an error if the target widget is null, or if the
     * reference widget is not null and not contained by the dock panel.
     */
    function validateInsertArgs(owner, widget, ref) {
        if (!widget) {
            throw new Error('Target widget is null.');
        }
        if (ref && !dockPanelContains(owner, ref)) {
            throw new Error('Reference widget not contained by the dock panel.');
        }
    }
    /**
     * Recursively iterate over the dock tab panels of a root panel.
     *
     * Iteration stops if the callback returns anything but `undefined`.
     */
    function iterTabPanels(root, callback) {
        if (root instanceof DockTabPanel) {
            return callback(root);
        }
        if (!(root instanceof DockSplitPanel)) {
            internalError();
        }
        for (var i = 0; i < root.childCount(); ++i) {
            var child = root.childAt(i);
            var result = iterTabPanels(child, callback);
            if (result !== void 0)
                return result;
        }
        return void 0;
    }
    /**
     * Get the root edge zone for the given node and client position.
     *
     * This assumes the position lies within the node's client rect.
     *
     * Returns the `Invalid` zone if the position is not within an edge.
     */
    function getEdgeZone(node, x, y) {
        var zone;
        var rect = node.getBoundingClientRect();
        if (x < rect.left + EDGE_SIZE) {
            if (y - rect.top < x - rect.left) {
                zone = 0 /* RootTop */;
            }
            else if (rect.bottom - y < x - rect.left) {
                zone = 3 /* RootBottom */;
            }
            else {
                zone = 1 /* RootLeft */;
            }
        }
        else if (x >= rect.right - EDGE_SIZE) {
            if (y - rect.top < rect.right - x) {
                zone = 0 /* RootTop */;
            }
            else if (rect.bottom - y < rect.right - x) {
                zone = 3 /* RootBottom */;
            }
            else {
                zone = 2 /* RootRight */;
            }
        }
        else if (y < rect.top + EDGE_SIZE) {
            zone = 0 /* RootTop */;
        }
        else if (y >= rect.bottom - EDGE_SIZE) {
            zone = 3 /* RootBottom */;
        }
        else {
            zone = 10 /* Invalid */;
        }
        return zone;
    }
    /**
     * Get the panel zone for the given node and position.
     *
     * This assumes the position lies within the node's client rect.
     *
     * This always returns a valid zone.
     */
    function getPanelZone(node, x, y) {
        var zone;
        var rect = node.getBoundingClientRect();
        var fracX = (x - rect.left) / rect.width;
        var fracY = (y - rect.top) / rect.height;
        if (fracX < 1 / 3) {
            if (fracY < fracX) {
                zone = 5 /* PanelTop */;
            }
            else if (1 - fracY < fracX) {
                zone = 8 /* PanelBottom */;
            }
            else {
                zone = 6 /* PanelLeft */;
            }
        }
        else if (fracX < 2 / 3) {
            if (fracY < 1 / 3) {
                zone = 5 /* PanelTop */;
            }
            else if (fracY < 2 / 3) {
                zone = 9 /* PanelCenter */;
            }
            else {
                zone = 8 /* PanelBottom */;
            }
        }
        else {
            if (fracY < 1 - fracX) {
                zone = 5 /* PanelTop */;
            }
            else if (fracY > fracX) {
                zone = 8 /* PanelBottom */;
            }
            else {
                zone = 7 /* PanelRight */;
            }
        }
        return zone;
    }
    /**
     * The current tab drag object.
     */
    var currentDrag = null;
    /**
     * Create a new tab panel for a dock panel.
     */
    function createTabPanel() {
        var panel = new DockTabPanel();
        panel.tabBar.tabDetachRequested.connect(onTabDetachRequested);
        panel.stackedPanel.widgetRemoved.connect(onWidgetRemoved);
        return panel;
    }
    /**
     * Remove an empty dock tab panel from the hierarchy.
     *
     * This ensures that the hierarchy is kept consistent by merging an
     * ancestor split panel when it contains only a single child widget.
     */
    function removeTabPanel(tabPanel) {
        // Assert the tab panel is empty.
        if (tabPanel.childCount() !== 0) {
            internalError();
        }
        // If the parent of the tab panel is a dock panel, just remove it.
        if (tabPanel.parent instanceof DockPanel) {
            setRoot(tabPanel.parent, null);
            tabPanel.dispose();
            return;
        }
        // Assert the tab panel parent is a dock split panel.
        if (!(tabPanel.parent instanceof DockSplitPanel)) {
            internalError();
        }
        // Cast the tab panel parent to a dock split panel.
        var splitPanel = tabPanel.parent;
        // Assert the split panel has at least two children.
        if (splitPanel.childCount() < 2) {
            internalError();
        }
        // Dispose the tab panel to ensure its resources are released.
        tabPanel.dispose();
        // If the split panel still has multiple children, there is
        // nothing more to do.
        if (splitPanel.childCount() > 1) {
            return;
        }
        // Extract the remaining child from the split panel.
        var child = splitPanel.childAt(0);
        // Assert the remaining child is a proper panel type.
        if (!(child instanceof DockTabPanel) && !(child instanceof DockSplitPanel)) {
            internalError();
        }
        // If the parent of the split panel is a dock panel, replace it.
        if (splitPanel.parent instanceof DockPanel) {
            setRoot(splitPanel.parent, child);
            splitPanel.dispose();
            return;
        }
        // Assert the split panel parent is a dock split panel.
        if (!(splitPanel.parent instanceof DockSplitPanel)) {
            internalError();
        }
        // Cast the split panel parent to a dock split panel.
        var grandPanel = splitPanel.parent;
        // If the child is a dock tab panel, replace the split panel.
        if (child instanceof DockTabPanel) {
            var sizes = grandPanel.sizes();
            var index_1 = grandPanel.childIndex(splitPanel);
            splitPanel.parent = null;
            grandPanel.insertChild(index_1, child);
            grandPanel.setSizes(sizes);
            splitPanel.dispose();
            return;
        }
        // Cast the child to a dock split panel.
        var childSplit = child;
        // Child splitters have an orthogonal orientation to their parent.
        // Assert the orientation of the child matches the grand parent.
        if (childSplit.orientation !== grandPanel.orientation) {
            internalError();
        }
        // The grand children can now be merged with their grand parent.
        // Start by fetching the relevant current sizes and insert index.
        var index = grandPanel.childIndex(splitPanel);
        var childSizes = childSplit.sizes();
        var grandSizes = grandPanel.sizes();
        // Remove the split panel and store its share of the size.
        splitPanel.parent = null;
        var sizeShare = arrays.removeAt(grandSizes, index);
        // Merge the grand children and maintain their relative size.
        for (var i = 0; childSplit.childCount() !== 0; ++i) {
            grandPanel.insertChild(index + i, childSplit.childAt(0));
            arrays.insert(grandSizes, index + i, sizeShare * childSizes[i]);
        }
        // Update the grand parent sizes and dispose the removed panel.
        grandPanel.setSizes(grandSizes);
        splitPanel.dispose();
    }
    /**
     * Handle the `tabDetachRequested` signal from a dock tab bar.
     */
    function onTabDetachRequested(sender, args) {
        // Do nothing if a drag is already in progress.
        if (currentDrag) {
            return;
        }
        // Release the tab bar's hold on the mouse.
        sender.releaseMouse();
        // Setup the mime data for the drag operation.
        var mimeData = new phosphor_dragdrop_1.MimeData();
        var widget = args.item;
        mimeData.setData(FACTORY_MIME, function () { return widget; });
        // Create the drag image for the drag operation.
        var tab = sender.tabAt(args.index);
        var dragImage = tab.cloneNode(true);
        // Create the drag object to manage the drag-drop operation.
        currentDrag = new phosphor_dragdrop_1.Drag({
            mimeData: mimeData,
            dragImage: dragImage,
            proposedAction: phosphor_dragdrop_1.DropAction.Move,
            supportedActions: phosphor_dragdrop_1.DropActions.Move,
        });
        // Start the drag operation and cleanup when done.
        tab.classList.add(HIDDEN_CLASS);
        currentDrag.start(args.clientX, args.clientY).then(function () {
            currentDrag = null;
            tab.classList.remove(HIDDEN_CLASS);
        });
    }
    /**
     * Handle the `widgetRemvoed` signal for a dock stacked panel.
     */
    function onWidgetRemoved(sender, widget) {
        if (sender.childCount() === 0) {
            removeTabPanel(sender.parent);
        }
    }
})(DockPanelPrivate || (DockPanelPrivate = {}));
