// phosphor-boxpanel/lib/layout.js
/*-----------------------------------------------------------------------------
| Copyright (c) 2014-2015, PhosphorJS Contributors
|
| Distributed under the terms of the BSD 3-Clause License.
|
| The full license is in the file LICENSE, distributed with this software.
|----------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var arrays = require('phosphor-arrays');
var phosphor_boxengine_1 = require('phosphor-boxengine');
var phosphor_domutil_1 = require('phosphor-domutil');
var phosphor_messaging_1 = require('phosphor-messaging');
var phosphor_properties_1 = require('phosphor-properties');
var phosphor_panel_1 = require('phosphor-panel');
var phosphor_widget_1 = require('phosphor-widget');
/**
 * The class name added to left-to-right box layout parents.
 */
var LEFT_TO_RIGHT_CLASS = 'p-mod-left-to-right';
/**
 * The class name added to right-to-left box layout parents.
 */
var RIGHT_TO_LEFT_CLASS = 'p-mod-right-to-left';
/**
 * The class name added to top-to-bottom box layout parents.
 */
var TOP_TO_BOTTOM_CLASS = 'p-mod-top-to-bottom';
/**
 * The class name added to bottom-to-top box layout parents.
 */
var BOTTOM_TO_TOP_CLASS = 'p-mod-bottom-to-top';
/**
 * The layout direction of a box layout.
 */
(function (Direction) {
    /**
     * Left to right direction.
     */
    Direction[Direction["LeftToRight"] = 0] = "LeftToRight";
    /**
     * Right to left direction.
     */
    Direction[Direction["RightToLeft"] = 1] = "RightToLeft";
    /**
     * Top to bottom direction.
     */
    Direction[Direction["TopToBottom"] = 2] = "TopToBottom";
    /**
     * Bottom to top direction.
     */
    Direction[Direction["BottomToTop"] = 3] = "BottomToTop";
})(exports.Direction || (exports.Direction = {}));
var Direction = exports.Direction;
/**
 * A layout which arranges its children in a single row or column.
 */
var BoxLayout = (function (_super) {
    __extends(BoxLayout, _super);
    function BoxLayout() {
        _super.apply(this, arguments);
        this._fixed = 0;
        this._spacing = 8;
        this._box = null;
        this._sizers = [];
        this._direction = Direction.TopToBottom;
    }
    Object.defineProperty(BoxLayout.prototype, "direction", {
        /**
         * Get the layout direction for the box layout.
         */
        get: function () {
            return this._direction;
        },
        /**
         * Set the layout direction for the box layout.
         */
        set: function (value) {
            if (this._direction === value) {
                return;
            }
            this._direction = value;
            if (!this.parent) {
                return;
            }
            BoxLayoutPrivate.toggleDirection(this.parent, value);
            this.parent.fit();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BoxLayout.prototype, "spacing", {
        /**
         * Get the inter-element spacing for the box layout.
         */
        get: function () {
            return this._spacing;
        },
        /**
         * Set the inter-element spacing for the box layout.
         */
        set: function (value) {
            value = Math.max(0, value | 0);
            if (this._spacing === value) {
                return;
            }
            this._spacing = value;
            if (!this.parent) {
                return;
            }
            this.parent.fit();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Initialize the children of the layout.
     *
     * #### Notes
     * This method is called automatically when the layout is installed
     * on its parent widget.
     */
    BoxLayout.prototype.initialize = function () {
        BoxLayoutPrivate.toggleDirection(this.parent, this.direction);
        _super.prototype.initialize.call(this);
    };
    /**
     * Attach a child widget to the parent's DOM node.
     *
     * @param index - The current index of the child in the layout.
     *
     * @param child - The child widget to attach to the parent.
     *
     * #### Notes
     * This is a reimplementation of the superclass method.
     */
    BoxLayout.prototype.attachChild = function (index, child) {
        arrays.insert(this._sizers, index, new phosphor_boxengine_1.BoxSizer());
        BoxLayoutPrivate.prepareGeometry(child);
        this.parent.node.appendChild(child.node);
        if (this.parent.isAttached)
            phosphor_messaging_1.sendMessage(child, phosphor_widget_1.Widget.MsgAfterAttach);
        this.parent.fit();
    };
    /**
     * Move a child widget in the parent's DOM node.
     *
     * @param fromIndex - The previous index of the child in the layout.
     *
     * @param toIndex - The current index of the child in the layout.
     *
     * @param child - The child widget to move in the parent.
     *
     * #### Notes
     * This is a reimplementation of the superclass method.
     */
    BoxLayout.prototype.moveChild = function (fromIndex, toIndex, child) {
        arrays.move(this._sizers, fromIndex, toIndex);
        this.parent.update();
    };
    /**
     * Detach a child widget from the parent's DOM node.
     *
     * @param index - The previous index of the child in the layout.
     *
     * @param child - The child widget to detach from the parent.
     *
     * #### Notes
     * This is a reimplementation of the superclass method.
     */
    BoxLayout.prototype.detachChild = function (index, child) {
        arrays.removeAt(this._sizers, index);
        if (this.parent.isAttached)
            phosphor_messaging_1.sendMessage(child, phosphor_widget_1.Widget.MsgBeforeDetach);
        this.parent.node.removeChild(child.node);
        BoxLayoutPrivate.resetGeometry(child);
        this.parent.fit();
    };
    /**
     * A message handler invoked on an `'after-show'` message.
     */
    BoxLayout.prototype.onAfterShow = function (msg) {
        _super.prototype.onAfterShow.call(this, msg);
        this.parent.update();
    };
    /**
     * A message handler invoked on an `'after-attach'` message.
     */
    BoxLayout.prototype.onAfterAttach = function (msg) {
        _super.prototype.onAfterAttach.call(this, msg);
        this.parent.fit();
    };
    /**
     * A message handler invoked on a `'child-shown'` message.
     */
    BoxLayout.prototype.onChildShown = function (msg) {
        if (BoxLayoutPrivate.IsIE) {
            phosphor_messaging_1.sendMessage(this.parent, phosphor_widget_1.Widget.MsgFitRequest);
        }
        else {
            this.parent.fit();
        }
    };
    /**
     * A message handler invoked on a `'child-hidden'` message.
     */
    BoxLayout.prototype.onChildHidden = function (msg) {
        if (BoxLayoutPrivate.IsIE) {
            phosphor_messaging_1.sendMessage(this.parent, phosphor_widget_1.Widget.MsgFitRequest);
        }
        else {
            this.parent.fit();
        }
    };
    /**
     * A message handler invoked on a `'resize'` message.
     */
    BoxLayout.prototype.onResize = function (msg) {
        if (this.parent.isVisible) {
            this._update(msg.width, msg.height);
        }
    };
    /**
     * A message handler invoked on an `'update-request'` message.
     */
    BoxLayout.prototype.onUpdateRequest = function (msg) {
        if (this.parent.isVisible) {
            this._update(-1, -1);
        }
    };
    /**
     * A message handler invoked on a `'fit-request'` message.
     */
    BoxLayout.prototype.onFitRequest = function (msg) {
        if (this.parent.isAttached) {
            this._fit();
        }
    };
    /**
     * Fit the layout to the total size required by the child widgets.
     */
    BoxLayout.prototype._fit = function () {
        // Compute the visible item count.
        var nVisible = 0;
        for (var i = 0, n = this.childCount(); i < n; ++i) {
            if (!this.childAt(i).isHidden)
                nVisible++;
        }
        // Update the fixed space for the visible items.
        this._fixed = this._spacing * Math.max(0, nVisible - 1);
        // Setup the initial size limits.
        var minW = 0;
        var minH = 0;
        var maxW = Infinity;
        var maxH = Infinity;
        var horz = BoxLayoutPrivate.isHorizontal(this._direction);
        if (horz) {
            minW = this._fixed;
            maxW = nVisible > 0 ? minW : maxW;
        }
        else {
            minH = this._fixed;
            maxH = nVisible > 0 ? minH : maxH;
        }
        // Update the sizers and computed size limits.
        for (var i = 0, n = this.childCount(); i < n; ++i) {
            var child = this.childAt(i);
            var sizer = this._sizers[i];
            if (child.isHidden) {
                sizer.minSize = 0;
                sizer.maxSize = 0;
                continue;
            }
            var limits = phosphor_domutil_1.sizeLimits(child.node);
            sizer.sizeHint = BoxLayout.getSizeBasis(child);
            sizer.stretch = BoxLayout.getStretch(child);
            if (horz) {
                sizer.minSize = limits.minWidth;
                sizer.maxSize = limits.maxWidth;
                minW += limits.minWidth;
                maxW += limits.maxWidth;
                minH = Math.max(minH, limits.minHeight);
                maxH = Math.min(maxH, limits.maxHeight);
            }
            else {
                sizer.minSize = limits.minHeight;
                sizer.maxSize = limits.maxHeight;
                minH += limits.minHeight;
                maxH += limits.maxHeight;
                minW = Math.max(minW, limits.minWidth);
                maxW = Math.min(maxW, limits.maxWidth);
            }
        }
        // Update the box sizing and add it to the size constraints.
        var box = this._box = phosphor_domutil_1.boxSizing(this.parent.node);
        minW += box.horizontalSum;
        minH += box.verticalSum;
        maxW += box.horizontalSum;
        maxH += box.verticalSum;
        // Update the parent's size constraints.
        var style = this.parent.node.style;
        style.minWidth = minW + "px";
        style.minHeight = minH + "px";
        style.maxWidth = maxW === Infinity ? 'none' : maxW + "px";
        style.maxHeight = maxH === Infinity ? 'none' : maxH + "px";
        // Notify the ancestor that it should fit immediately.
        var ancestor = this.parent.parent;
        if (ancestor)
            phosphor_messaging_1.sendMessage(ancestor, phosphor_widget_1.Widget.MsgFitRequest);
        // Notify the parent that it should update immediately.
        phosphor_messaging_1.sendMessage(this.parent, phosphor_widget_1.Widget.MsgUpdateRequest);
    };
    /**
     * Update the layout position and size of the child widgets.
     *
     * The parent offset dimensions should be `-1` if unknown.
     */
    BoxLayout.prototype._update = function (offsetWidth, offsetHeight) {
        // Bail early if there are no children to layout.
        if (this.childCount() === 0) {
            return;
        }
        // Measure the parent if the offset dimensions are unknown.
        if (offsetWidth < 0) {
            offsetWidth = this.parent.node.offsetWidth;
        }
        if (offsetHeight < 0) {
            offsetHeight = this.parent.node.offsetHeight;
        }
        // Ensure the parent box sizing data is computed.
        var box = this._box || (this._box = phosphor_domutil_1.boxSizing(this.parent.node));
        // Compute the layout area adjusted for border and padding.
        var top = box.paddingTop;
        var left = box.paddingLeft;
        var width = offsetWidth - box.horizontalSum;
        var height = offsetHeight - box.verticalSum;
        // Distribute the layout space and adjust the start position.
        switch (this._direction) {
            case Direction.LeftToRight:
                phosphor_boxengine_1.boxCalc(this._sizers, Math.max(0, width - this._fixed));
                break;
            case Direction.TopToBottom:
                phosphor_boxengine_1.boxCalc(this._sizers, Math.max(0, height - this._fixed));
                break;
            case Direction.RightToLeft:
                phosphor_boxengine_1.boxCalc(this._sizers, Math.max(0, width - this._fixed));
                left += width;
                break;
            case Direction.BottomToTop:
                phosphor_boxengine_1.boxCalc(this._sizers, Math.max(0, height - this._fixed));
                top += height;
                break;
        }
        // Layout the children using the computed box sizes.
        for (var i = 0, n = this.childCount(); i < n; ++i) {
            var child = this.childAt(i);
            if (child.isHidden) {
                continue;
            }
            var size = this._sizers[i].size;
            switch (this._direction) {
                case Direction.LeftToRight:
                    BoxLayoutPrivate.setGeometry(child, left, top, size, height);
                    left += size + this._spacing;
                    break;
                case Direction.TopToBottom:
                    BoxLayoutPrivate.setGeometry(child, left, top, width, size);
                    top += size + this._spacing;
                    break;
                case Direction.RightToLeft:
                    BoxLayoutPrivate.setGeometry(child, left - size, top, size, height);
                    left -= size + this._spacing;
                    break;
                case Direction.BottomToTop:
                    BoxLayoutPrivate.setGeometry(child, left, top - size, width, size);
                    top -= size + this._spacing;
                    break;
            }
        }
    };
    return BoxLayout;
})(phosphor_panel_1.PanelLayout);
exports.BoxLayout = BoxLayout;
/**
 * The namespace for the `BoxLayout` class statics.
 */
var BoxLayout;
(function (BoxLayout) {
    /**
     * A convenience alias of the `LeftToRight` [[Direction]].
     */
    BoxLayout.LeftToRight = Direction.LeftToRight;
    /**
     * A convenience alias of the `RightToLeft` [[Direction]].
     */
    BoxLayout.RightToLeft = Direction.RightToLeft;
    /**
     * A convenience alias of the `TopToBottom` [[Direction]].
     */
    BoxLayout.TopToBottom = Direction.TopToBottom;
    /**
     * A convenience alias of the `BottomToTop` [[Direction]].
     */
    BoxLayout.BottomToTop = Direction.BottomToTop;
    /**
     * Get the box layout stretch factor for the given widget.
     *
     * @param widget - The widget of interest.
     *
     * @returns The box layout stretch factor for the widget.
     */
    function getStretch(widget) {
        return BoxLayoutPrivate.stretchProperty.get(widget);
    }
    BoxLayout.getStretch = getStretch;
    /**
     * Set the box layout stretch factor for the given widget.
     *
     * @param widget - The widget of interest.
     *
     * @param value - The value for the stretch factor.
     */
    function setStretch(widget, value) {
        BoxLayoutPrivate.stretchProperty.set(widget, value);
    }
    BoxLayout.setStretch = setStretch;
    /**
     * Get the box layout size basis for the given widget.
     *
     * @param widget - The widget of interest.
     *
     * @returns The box layout size basis for the widget.
     */
    function getSizeBasis(widget) {
        return BoxLayoutPrivate.sizeBasisProperty.get(widget);
    }
    BoxLayout.getSizeBasis = getSizeBasis;
    /**
     * Set the box layout size basis for the given widget.
     *
     * @param widget - The widget of interest.
     *
     * @param value - The value for the size basis.
     */
    function setSizeBasis(widget, value) {
        BoxLayoutPrivate.sizeBasisProperty.set(widget, value);
    }
    BoxLayout.setSizeBasis = setSizeBasis;
})(BoxLayout = exports.BoxLayout || (exports.BoxLayout = {}));
/**
 * The namespace for the `BoxLayout` class private data.
 */
var BoxLayoutPrivate;
(function (BoxLayoutPrivate) {
    /**
     * A flag indicating whether the browser is IE.
     */
    BoxLayoutPrivate.IsIE = /Trident/.test(navigator.userAgent);
    /**
     * The property descriptor for a widget stretch factor.
     */
    BoxLayoutPrivate.stretchProperty = new phosphor_properties_1.Property({
        name: 'stretch',
        value: 0,
        coerce: function (owner, value) { return Math.max(0, value | 0); },
        changed: onChildPropertyChanged,
    });
    /**
     * The property descriptor for a widget size basis.
     */
    BoxLayoutPrivate.sizeBasisProperty = new phosphor_properties_1.Property({
        name: 'sizeBasis',
        value: 0,
        coerce: function (owner, value) { return Math.max(0, value | 0); },
        changed: onChildPropertyChanged,
    });
    /**
     * Test whether a direction has horizontal orientation.
     */
    function isHorizontal(dir) {
        return dir === Direction.LeftToRight || dir === Direction.RightToLeft;
    }
    BoxLayoutPrivate.isHorizontal = isHorizontal;
    /**
     * Toggle the CSS direction class for the given widget.
     */
    function toggleDirection(widget, dir) {
        widget.toggleClass(LEFT_TO_RIGHT_CLASS, dir === Direction.LeftToRight);
        widget.toggleClass(RIGHT_TO_LEFT_CLASS, dir === Direction.RightToLeft);
        widget.toggleClass(TOP_TO_BOTTOM_CLASS, dir === Direction.TopToBottom);
        widget.toggleClass(BOTTOM_TO_TOP_CLASS, dir === Direction.BottomToTop);
    }
    BoxLayoutPrivate.toggleDirection = toggleDirection;
    /**
     * Prepare a child widget for absolute layout geometry.
     */
    function prepareGeometry(widget) {
        widget.node.style.position = 'absolute';
    }
    BoxLayoutPrivate.prepareGeometry = prepareGeometry;
    /**
     * Reset the layout geometry of a child widget.
     */
    function resetGeometry(widget) {
        var rect = rectProperty.get(widget);
        var style = widget.node.style;
        rect.top = NaN;
        rect.left = NaN;
        rect.width = NaN;
        rect.height = NaN;
        style.position = '';
        style.top = '';
        style.left = '';
        style.width = '';
        style.height = '';
    }
    BoxLayoutPrivate.resetGeometry = resetGeometry;
    /**
     * Set the layout geometry of a child widget.
     */
    function setGeometry(widget, left, top, width, height) {
        var resized = false;
        var style = widget.node.style;
        var rect = rectProperty.get(widget);
        if (rect.top !== top) {
            rect.top = top;
            style.top = top + "px";
        }
        if (rect.left !== left) {
            rect.left = left;
            style.left = left + "px";
        }
        if (rect.width !== width) {
            resized = true;
            rect.width = width;
            style.width = width + "px";
        }
        if (rect.height !== height) {
            resized = true;
            rect.height = height;
            style.height = height + "px";
        }
        if (resized) {
            phosphor_messaging_1.sendMessage(widget, new phosphor_widget_1.ResizeMessage(width, height));
        }
    }
    BoxLayoutPrivate.setGeometry = setGeometry;
    /**
     * A property descriptor for a widget offset rect.
     */
    var rectProperty = new phosphor_properties_1.Property({
        name: 'rect',
        create: function () { return ({ top: NaN, left: NaN, width: NaN, height: NaN }); },
    });
    /**
     * The change handler for the attached child properties.
     */
    function onChildPropertyChanged(child) {
        var parent = child.parent;
        var layout = parent && parent.layout;
        if (layout instanceof BoxLayout)
            parent.fit();
    }
})(BoxLayoutPrivate || (BoxLayoutPrivate = {}));
